frontends/concrete-python/concrete/fhe/compilation/decorators.py

```diff
@@ -75,92 +76,125 @@ def decoration(function: Callable):
+    """
+
+    function: Callable
+    compiler: Compiler
+
+    def __init__(self, function: Callable, parameters):
+        self.function = function  # type: ignore
+        self.compiler = Compiler(self.function, dict(parameters))
+
+    def __call__(self, *args, **kwargs) -> Any:
+        self.compiler(*args, **kwargs)
+        return self.function(*args, **kwargs)
+
+    def trace(
+        self,
+        inputset: Optional[Union[Iterable[Any], Iterable[Tuple[Any, ...]]]] = None,
+        configuration: Optional[Configuration] = None,
+        artifacts: Optional[DebugArtifacts] = None,
+        **kwargs,
+    ) -> Graph:
+        """
+        Trace the function into computation graph.
+
+        Args:
+            inputset (Optional[Union[Iterable[Any], Iterable[Tuple[Any, ...]]]]):
+                optional inputset to extend accumulated inputset before bounds measurement
+
+            configuration(Optional[Configuration], default = None):
+                configuration to use
+
+            artifacts (Optional[DebugArtifacts], default = None):
+                artifacts to store information about the process
+
+            kwargs (Dict[str, Any]):
+                configuration options to overwrite
+
+        Returns:
+            Graph:
+                computation graph representing the function prior to MLIR conversion
+        """
+
+        return self.compiler.trace(inputset, configuration, artifacts, **kwargs)
+
+    def compile(
+        self,
+        inputset: Optional[Union[Iterable[Any], Iterable[Tuple[Any, ...]]]] = None,
+        configuration: Optional[Configuration] = None,
+        artifacts: Optional[DebugArtifacts] = None,
+        **kwargs,
+    ) -> Circuit:
+        """
+        Compile the function into a circuit.
+
+        Args:
+            inputset (Optional[Union[Iterable[Any], Iterable[Tuple[Any, ...]]]]):
+                optional inputset to extend accumulated inputset before bounds measurement
+
+            configuration(Optional[Configuration], default = None):
+                configuration to use
+
+            artifacts (Optional[DebugArtifacts], default = None):
+                artifacts to store information about the process
+
+            kwargs (Dict[str, Any]):
+                configuration options to overwrite
+
+        Returns:
+            Circuit:
+                compiled circuit
+        """
+
+        return self.compiler.compile(inputset, configuration, artifacts, **kwargs)
+
+
 def compiler(parameters: Mapping[str, Union[str, EncryptionStatus]]):
     """
-    Provide an easy interface for compilation.
+    Provide an easy interface for the compilation of single-circuit programs.
+
+    Args:
+        parameters (Mapping[str, Union[str, EncryptionStatus]]):
+            encryption statuses of the parameters of the function to compile
+    """
+
+    def decoration(function: Callable):
+        return Compilable(function, parameters)
+
+    return decoration
+
+
+def program():
```

Comment:
is it to reserve any futur option ?

Reply:
What do you mean by reserving any futur option ?

Reply:
why not parameters ? (options == no parameter) you don't need that extra level if no parameters.
@fhe.program()
class Bla:
vs
@fhe.program
class Bla